src/frontend/src/optimizer/plan_node/stream_gap_fill.rs

```diff
@@ -0,0 +1,174 @@
+
+use risingwave_common::catalog::Field;
+use risingwave_common::types::DataType;
+use risingwave_common::util::sort_util::OrderType;
+use risingwave_pb::stream_plan::stream_node::NodeBody;
+
+use super::generic::GenericPlanNode;
+use super::utils::TableCatalogBuilder;
+use super::{
+    ExprRewritable, ExprVisitable, PlanBase, PlanRef, PlanTreeNodeUnary, Stream, StreamNode,
+    generic,
+};
+use crate::binder::BoundFillStrategy;
+use crate::expr::{Expr, ExprImpl, ExprRewriter, ExprVisitor, InputRef};
+use crate::optimizer::plan_node::stream::StreamPlanNodeMetadata;
+use crate::optimizer::plan_node::utils::impl_distill_by_unit;
+use crate::optimizer::property::Distribution;
+use crate::stream_fragmenter::BuildFragmentGraphState;
+
+/// `StreamGapFill` implements [`super::Stream`] to represent a gap-filling operation on a time
+/// series in normal streaming mode (without EOWC semantics).
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct StreamGapFill {
+    pub base: PlanBase<super::Stream>,
+    core: generic::GapFill<PlanRef<Stream>>,
+}
+
+impl StreamGapFill {
+    pub fn new(core: generic::GapFill<PlanRef<Stream>>) -> Self {
+        let input = &core.input;
+
+        // Use singleton distribution for normal streaming GapFill.
+        // Similar to EOWC version, gap filling requires seeing all data
+        // to correctly identify and fill gaps across time series.
+        let base = PlanBase::new_stream_with_core(
+            &core,
+            Distribution::Single,
+            input.stream_kind(),
+            false, // does NOT provide EOWC semantics
+            input.watermark_columns().clone(),
+            input.columns_monotonicity().clone(),
+        );
+        Self { base, core }
+    }
+
+    // Legacy constructor for backward compatibility
+    pub fn new_with_args(
+        input: PlanRef<Stream>,
+        time_col: InputRef,
+        interval: ExprImpl,
+        fill_strategies: Vec<BoundFillStrategy>,
+    ) -> Self {
+        let core = generic::GapFill {
+            input,
+            time_col,
+            interval,
+            fill_strategies,
+        };
+        Self::new(core)
+    }
+
+    pub fn time_col(&self) -> &InputRef {
+        &self.core.time_col
+    }
+
+    pub fn interval(&self) -> &ExprImpl {
+        &self.core.interval
+    }
+
+    pub fn fill_strategies(&self) -> &[BoundFillStrategy] {
+        &self.core.fill_strategies
+    }
+
+    fn infer_state_table(&self) -> crate::TableCatalog {
+        let mut tbl_builder = TableCatalogBuilder::default();
+
+        let out_schema = self.core.schema();
+        for field in out_schema.fields() {
+            tbl_builder.add_column(field);
+        }
+
+        // For singleton distribution, use simplified primary key design:
+        // Just use time column as the primary key for ordering
+        let time_col_idx = self.time_col().index();
+        tbl_builder.add_order_column(time_col_idx, OrderType::ascending());
```

Comment:
We'd better verify that `time_col` is indeed the pk of the upstream table. Otherwise if two rows correspond to the same PK we will have logic errors.